xbmc/filesystem/SpecialProtocol.cpp

```diff
@@ -28,12 +28,402 @@
+  return path;
+}
+
+std::string CSpecialProtocol::GetXBMCAltBinAddonPath(const std::string &fileName)
+{
+  return URIUtils::AddFileToFolder(GetXBMCAltBinAddonPath(), fileName);
+}
+
+
+std::string CSpecialProtocol::GetHomePath() 
+{
+  std::string path;
+#ifdef TARGET_WINDOWS
+  path = CWIN32Util::GetProfilePath();
+#endif
+#if defined(TARGET_POSIX) && !defined(TARGET_DARWIN)
+  if (g_application.PlatformDirectoriesEnabled()) 
+  {
+    std::string userHome;
+    std::string appName = CCompileInfo::GetAppName();
+    std::string dotLowerAppName = "." + appName;
+    StringUtils::ToLower(dotLowerAppName);
+    // REVIEW: Make a get function for userHome calc
+    if (getenv("HOME"))
+      userHome = getenv("HOME");
+    else
+      userHome = "/root";
+    path = userHome + "/" + dotLowerAppName;
+  }
+  else
+  {
+    path = URIUtils::AddFileToFolder(GetXBMCPath(), "portable_data");
+  }
+#endif
+#if defined(TARGET_DARWIN)
+  if (g_application.PlatformDirectoriesEnabled())
+  {
+    // REVIEW: Make a get function for userHome calc
+    std::string userHome;
+    if (getenv("HOME"))
+      userHome = getenv("HOME");
+    else
+      userHome = "/root";
+#if defined(TARGET_DARWIN_IOS)
+    path = userHome + "/" + CDarwinUtils::GetAppRootFolder() + "/" + CCompileInfo::GetAppName();
+#else
+    path = userHome + "/Library/Application Support/" + CCompileInfo::GetAppName();
+#endif
+  }
+  else
+  {
+    path = URIUtils::AddFileToFolder(GetXBMCPath(), "portable_data");
+  }
+#endif
+  return path;
+}
+
+std::string CSpecialProtocol::GetHomePath(const std::string &fileName) 
+{
+  return URIUtils::AddFileToFolder(GetHomePath(), fileName);
+}
+
+std::string CSpecialProtocol::GetEnvHomePath()
+{
+  std::string path;
+#if defined(TARGET_POSIX)
+  path =  getenv("HOME");
+#endif
+  return path;
+}
+
+std::string CSpecialProtocol::GetEnvHomePath(const std::string &fileName)
+{
+  return URIUtils::AddFileToFolder(GetEnvHomePath(), fileName);
+}
+
+
+std::string CSpecialProtocol::GetTmpPath(const std::string &fileName)
+{
+  return URIUtils::AddFileToFolder(GetTmpPath(), fileName);
+}
+
+std::string CSpecialProtocol::GetTmpPath() 
+{
+  std::string path;
+#ifdef TARGET_WINDOWS
+  path = URIUtils::AddFileToFolder(CWIN32Util::GetProfilePath(), "cache");
+#endif
+#if defined(TARGET_POSIX) && !defined(TARGET_DARWIN)
+  const char* envAppTemp = "KODI_TEMP";
+  // REVIEW: Make a get function for userHome calc
+  std::string userHome;
+  if (getenv("HOME"))
+    userHome = getenv("HOME");
+  else
+    userHome = "/root";
+
+  if (g_application.PlatformDirectoriesEnabled())
+  {
+    std::string strTempPath = userHome;
+    std::string appName = CCompileInfo::GetAppName();
+    // REVIEW: Make a get function for dotLowerAppName calc
+    std::string dotLowerAppName = "." + appName;
+    StringUtils::ToLower(dotLowerAppName);
+    strTempPath = URIUtils::AddFileToFolder(strTempPath, dotLowerAppName + "/temp");
```

Comment:
only use "temp", the slash is redundant and you can call this as 
`strTempPath = URIUtils::AddFileToFolder(strTempPath, dotLowerAppName, "temp");` it takes an "unlimited" amount of arguments.